spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1216 @@
+> current implementation of Fastsync (V2) requires the much
+> stronger requirement [**[FS-ALL-CORR-PEER]**](#FS-ALL-CORR-PEER)
+> given in Part V.
+
+### Safety
+
+> As this specification does
+> not assume that a correct peer is at the most recent height
+> of the blockchain (it might lag behind), the property [FS-SEQ-SAFE-START]
+> cannot be ensured in an unreliable distributed setting. We consider
+> the following relaxation. (Which is typically sufficient for
+> Tendermint, as the consensus reactor then synchronizes from that
+> height.)
+
+#### **[FS-DIST-SAFE-START]**
+
+Let *maxh* be the maximum
+height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
+
+> To address [FS-SEQ-SAFE-SYNC] we consider the following property in
+> the distributed setting. See the comments below on the relation to
+> the sequential version.
+
+#### **[FS-DIST-SAFE-SYNC]**
+
+Under [FS-SOME-CORR-PEER], there exists a constant time interval *TD*, such
+that if *term* is the time *Fastsync* terminates and
+*maxh* is the maximum height of a correct peer
+[**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link] in *peerIDs* at the time
+*term - TD*, then if *FastSync* terminates successfully, it is at
+some height *terminationHeight >= maxh - 1*.
+
+> *TD* might depend on timeouts etc. We suggest that an acceptable
+> value for *TD* is in the range of approx. 10 sec., that is the
+> interval between two calls `QueryStatus()`; see below.
+> We use *term - TD* as reference time, as we have to account
+> for communication delay between the peer and *FS*. After the peer sent
+> the last message to *FS*, the peer and *FS* run concurrently and
+> independently. There is no assumption on the rate at which a peer can
+> add blocks (e.g., it might be in the process of catching up
+> itself). Hence, without additional assumption we cannot link
+> [FS-DIST-SAFE-SYNC] to
+> [**[FS-SEQ-SAFE-SYNC]**](#FS-SEQ-SAFE-SYNC), in particular to the
+> parameter *D*. We discuss a
+> way to achieve this below:
+> **Relation to [FS-SEQ-SAFE-SYNC]:**  
+> Under [FS-SOME-CORR-PEER], if *peerIDs* contains a full node that is
+> "synchronized with the blockchain", and *blockchainheight* is the height
+> of the blockchain at time *term*, then  *terminationHeight* may even
+> achieve
+> *blockchainheight - TD / ETIME*;
+> cf. [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link], that is,
+> the parameter *D* from [FS-SEQ-SAFE-SYNC] is in the range of  *TD / ETIME*.
+
+#### **[FS-DIST-SAFE-STATE]**
+
+It is the same as the sequential version
+[**[FS-SEQ-SAFE-STATE]**](#FS-SEQ-SAFE-STATE).
+
+#### **[FS-DIST-NONABORT]**
+
+If there is one correct process in *peerIDs* [FS-SOME-CORR-PEER],
+*Fastsync* never terminates with failure. (Together with [FS-DIST-LIVE]
+ that means it will terminate successfully.)
+
+# Part IV - Fastsync protocol
```

Comment:
This sections is specific to the go implementation or will the rust implementation be following the same spec?